tortoise/verifying_algorithm.go

```diff
@@ -39,14 +33,32 @@ type Config struct {
 	RerunInterval   time.Duration // how often to rerun from genesis
 }
 
+// ThreadSafeVerifyingTortoise is a thread safe verifying tortoise wrapper, it just locks all actions.
+type ThreadSafeVerifyingTortoise struct {
+	// update for rerun
+	update atomic.UnsafePointer
+
+	logger log.Log
+
+	eg     errgroup.Group
+	cancel context.CancelFunc
+
+	mu sync.RWMutex
+	// persistMu is needed to allow concurrent BaseBlock and Persist call
+	// but to prevent multiple concurrent Persist calls
+	// relevant after rerun on nodes with slow disk
+	persistMu sync.Mutex
```

Comment:
oh. i guess you just wanted to have one `Persist()` at a time. 

intuitively the locking scheme is complicated by the fact that `Persist()` is called by the mesh. it seems more natural to me to call `state.persist()` internally inside tortoise. that way tortoise have more control of persisting atomic state. as the code stands now, we could potentially process half of the blocks in a layer N and abort, leaving half of the layer N votes in memory, and then they get persisted as a result of layer N+1.

Reply:
i would rather remove state completely eventually. 

as for the issue you pointed out i don't see how moving persist to tortoise will change anything. if we can abort with half-computed state it will be anyway persisted by the next call.

Reply:
> oh. i guess you just wanted to have one Persist() at a time.

yes, but it is important to use only mu.RLock so that BaseBlock is never blocked by Persist

Reply:
> as for the issue you pointed out i don't see how moving persist to tortoise will change anything. if we can abort with half-computed state it will be anyway persisted by the next call.

it won't make a difference now. but in the future, if persist() is called inside of tortoise only, we know exactly the place we aborted and revert the update in memory.

how do you envision removing state completely in the future? how would we store the opinions?

Reply:
i meant removing the persistent state. we can always recover votes from blocks when application starts.

we may do it even now, currently, we only need votes from the sliding window, and as it was discussed in slack we may even change that window so that we only need votes that start from the verified layer.

we don't even need to do a full rerun when the application starts, just a rerun starting from verified - N layer. i want to double-check if i am not missing anything, but it will be great if the persistent state can be removed.

Reply:
i agree with everything with a slight modification.

> we don't even need to do a full rerun when the application starts, just a rerun starting from verified - N layer. i want to double-check if i am not missing anything, but it will be great if the persistent state can be removed.

like i said in the slack also, `verified` is a moving target. if we find that our mesh hash differ wither peers further back than `verified`, we need to shift the sliding window back where `verified` is the point of convergence in mesh hash.

Reply:
i think i understand it differently. if we rerun from scratch it means that our sliding window starts from genesis. if hash differs at some layer we can start sliding window from that layer and rerun from that layer. i think the concept of sliding window was introduced because by design we had to maintain some state in memory for self-healing. if we don't need to maintain it in memory then there is no such thing as a sliding window.

there is another use case for keeping votes in memory, when the verified layer is stuck we want to keep all the votes so that it can unstuck eventually. this is what i was arguing about in the slack.

Reply:
currently the only way to trigger self-healing is `verified layer is stuck`. so i think these two scenarios are similar.

> if hash differs at some layer we can start sliding window from that layer and rerun from that layer. 

that's effectively the same as what i meant. say our hash differ from peers starting from layer N, current verified layer is N+1000. we need to move verified layer back to N and rerun from there.